### DISCUSSION TITLE
Performance on macOS vs Windows (.net v6 console application, C# v9)

### DISCUSSION THREAD

**Discussion by PlunderBunny** [dotnet/runtime#76797] [Q&A]
Hi all.

I have run up against an interesting performance issue when comparing the exact same .net code in Windows and macOS. I don't understand why there's such a significant performance difference, and I'm not sure of the best way to proceed.

The code is for a .net (v6) console application (C# v9) that I've been developing on macOS using Visual Studio for Mac. The program performs acceptably on macOS. When I copy the code to my Windows VM and re-compile in Visual Studio 2022, I find that the performance is quite poor - unusably so.

I did some simple performance investigations using the Stopwatch class, beginning with the method that writes to the console window: I determined this was more than 150x slower on Windows than macOS - please see [my SO post](https://stackoverflow.com/questions/73945754/performance-difference-between-windows-and-macos-net-console-app/) that includes a minimal reproducible example and more detail.

I then looked at some other parts of the application, and it appears that performance under Windows is slower than macOS even for code that consists of in-memory (i.e. no I/O) calculations only. For example, my routine to determine what the player can see executes at-least 10x slower on Windows than on macOS. I emphasise that this is exactly the same code, with both versions built on up-to-date Microsoft compilers.

It's as-if the Windows version of the application is being throttled (I don't mean that in a nefarious way). A release version behaves similarly. Is this a recognised issue, and if so, is there a work-around? I know almost nothing about the more technical aspects of the .net runtime, so I'd really appreciate a simplified explanation.

Thanks for any help.

**Comment by Fabi** [comment]
Console output benchmarks? Doubt that this is any important thing. Filling a console is NOT usable and does not make sense for a benchmark at all.

**Reply by PlunderBunny** [reply]
Hi Fabi.
I agree that the minimal reproducible example (filling the console window) isn't very representative, and by itself doesn't necessarily mean anything with respect to the performance of the .net runtime on Windows vs macOS. I can optimise the console output in the real application to make it insignificant. But nevertheless, I see the trend across the entire application, and the result is that the macOS version is usable, but the Windows version (with exactly the same code) is unusably slow, and that's what I'm trying to understand.

**Comment by LeaFrock** [comment]
Do you actually mean `Performance on macOS vs Windows Virtual Machine`? That sounds reasonable.

**Reply by PlunderBunny** [reply]
I've tested it on both a Windows VM and real hardware, and the performance difference is present on both. My Windows VM is hosted on real hardware, and I've never noticed a significant difference between it and a real Windows box (I use it for my day job, and run Visual Studio, Office apps etc and they all perform fine). I don't see why a VM would be significantly different from real hardware, provided it's not resource constrained (which, in my case, it's not).
Nevertheless, I'll try again when I get a chance.

**Reply by danmoseley** [reply]
Can you provide standalone repro code that is significantly slower on Mac? As folks have said, console on Unix can be quite different but you mentioned also code that doesn't use console.

What settings do you have in your project file/pass as build flags?

**Reply by PlunderBunny** [reply]
Re: Console output example, here's the minimal reproducible example from [my SO post](https://stackoverflow.com/questions/73945754/performance-difference-between-windows-and-macos-net-console-app/), simplified to output milliseconds. On macOS in Visual Studio for Mac (v17.3.6), it consistently outputs 2ms. On Windows 10 in Visual Studio 2022 (v17.1.6), it displays values between 220ms and 240ms, or if I use the legacy console mode, 280ms - 300ms. Both in debug mode. Project/solution settings are the default for a .net 6.0 console app. If you have a real version of Windows (as opposed to a VM like me) perhaps you could compare the timings you get running the same code?
(Your post said "...significantly slower on Mac..." but the issue is that the code is significantly slower on Windows. I assume that's what you meant).

I'll try to make a simple example of code that doesn't output anything.
```
using System.Diagnostics;

Stopwatch stopwatch = new();

while (true)
{
	stopwatch.Restart();

	for (int y = 0; y < Console.WindowHeight; ++y)
	{
		Console.SetCursorPosition(0, y);
		for (int x = 0; x < Console.WindowWidth; ++x)
		{
			Console.Write('A');
		}
	}

	stopwatch.Stop();
	Debug.WriteLine($"{stopwatch.ElapsedMilliseconds}ms");
}
```

**Reply by danmoseley** [reply]
Thanks for the minimal repro. I suggest you now open an issue in this repo for console folks to look at.
Unfortunately I don't have a Mac to try, but they do.

> I assume that's what you meant

yes my bad

**Comment by hez2010** [comment]
You will also find the performance can be largely different among different terminals even in the same OS.

**Reply by PlunderBunny** [reply]
Good point. Under Windows, is there any way to make a .net console app output to something other than the command-line console, e.g. to a powershell console (I'm probably mixing up my terminology)?
(For what it's worth, I'm not using the legacy console mode, although I have tried it and if anything it seems to make the problem worse - the minimal reproducible example runs slightly slower in legacy mode)

**Reply by hez2010** [reply]
Try the new Windows Terminal.

**Reply by PlunderBunny** [reply]
Thanks. I have installed the new Windows Terminal, and tried running a release version of my application from a command-prompt tab within the terminal application. The performance is no different. I suspect that the new Windows Terminal just hosts the existing command prompt, so I'm not surprised it isn't any better.
